pulsar-broker/src/test/java/org/apache/pulsar/client/api/InjectedErrorProducerConsumerTest.java

```diff
@@ -0,0 +1,184 @@
+        admin.topics().createNonPartitionedTopic(topicName);
+        Consumer<String> consumer = pulsarClient.newConsumer(Schema.STRING).topic(topicName)
+                .subscriptionName(subscribe).receiverQueueSize(5)
+                .subscriptionType(subscriptionType).subscribe();
+        consumer.close();
+        PersistentTopic persistentTopic =
+                (PersistentTopic) pulsar.getBrokerService().getTopic(topicName, true).join().get();
+        PersistentSubscription subscription = persistentTopic.getSubscription(subscribe);
+        InjectedSubscription injectedSubscription = new InjectedSubscription(persistentTopic, subscription);
+        injectedSubscription.spySubscription();
+        injectedSubscription.spyDispatcher();
+        return injectedSubscription;
+    }
+
+    @Data
+    @AllArgsConstructor
+    private class InjectedSubscription {
+        private PersistentTopic persistentTopic;
+        private PersistentSubscription subscription;
+        private PersistentSubscription spySubscription;
+        private Dispatcher dispatcher;
+        private Dispatcher spyDispatcher;
+        private AtomicInteger subscribeCounter = new AtomicInteger();
+        private HashMap<Integer, CountDownLatch> whichSubscribeTimeout = new HashMap<>();
+
+        private InjectedSubscription(PersistentTopic persistentTopic, PersistentSubscription subscription) {
+            this.persistentTopic = persistentTopic;
+            this.subscription = subscription;
+        }
+
+        private void spySubscription() throws Exception {
+            spySubscription = spy(subscription);
+            persistentTopic.getSubscriptions().put(subscription.getName(), spySubscription);
+            doAnswer(invocation -> {
+                int times = subscribeCounter.incrementAndGet();
+                org.apache.pulsar.broker.service.Consumer consumer =
+                        (org.apache.pulsar.broker.service.Consumer) invocation.getArguments()[0];
+                CompletableFuture<Void> future = subscription.addConsumer(consumer);
+                return future.thenAccept(ignore -> {
+                    CountDownLatch signal = whichSubscribeTimeout.get(times);
+                    if (signal != null) {
+                        sneakyAwait(signal);
+                    }
+                });
+            }).when(spySubscription).addConsumer(any(org.apache.pulsar.broker.service.Consumer.class));
+        }
+
+        private boolean spyDispatcher() throws Exception {
+            dispatcher = subscription.getDispatcher();
+            if (dispatcher == null) {
+                return false;
+            }
+            spyDispatcher = spy(dispatcher);
+            Field fieldDispatcher = PersistentSubscription.class.getDeclaredField("dispatcher");
+            fieldDispatcher.setAccessible(true);
+            fieldDispatcher.set(subscription, spyDispatcher);
+            return true;
+        }
```

Comment:
Why did you spy the dispatcher? It seems that `spyDispatcher` does not have different behavior with the original dispatcher.

Reply:
It doesn't affect the result. You can remove this `spy`